### PR TITLE
Mpfs icicle qspi support

### PR DESCRIPTION
**Pull request by con-pax**
This patch adds driver support for Microchip's Polarfire SOC MSS QSPI peripheral.
Tested this patch on Microchip's ICICLE kit which has MT25QL256 Micron SPI NOR
flash on it. This is tested using generic spi-nor flash driver.
This patch also adds the spi-nor flash node to mpfs-icicle.dts.

### CODE REVIEW

**Comment by fkokosinski** [comment]
Hi @con-pax! Thanks for your contribution.

It looks like a lot of build failures in the CI can be linked to the `mss_qspi*` symbols being declared, but not being used anywhere. Could you fix this?

**Comment by con-pax** [comment]
> Hi @con-pax! Thanks for your contribution.
> 
> It looks like a lot of build failures in the CI can be linked to the `mss_qspi*` symbols being declared, but not being used anywhere. Could you fix this?

Hi @fkokosinski, Thanks for checking this out. Yes, we are in the process of fixing and rebasing.

**Comment by fkokosinski** [comment]
> Can we ignore this error? As this warning is treated as error in this test. The other SPI drivers are also modeled in the same manner. Can you please let me know, if I need to correct something here?
> 
> Thanks, Suresh.

Sorry, unfortunately this PR won't be merged if CI isn't green. Are you sure that the `MSS_QSPI_INIT` is used at all? I.e., how does this file look like after preprocessing. Because it looks like the `DT_INST_FOREACH_STATUS_OKAY` doesn't call it at all, and that's why we get these unused symbols. Could you check that?

**Comment by nagasureshkumar** [comment]
> > Can we ignore this error? As this warning is treated as error in this test. The other SPI drivers are also modeled in the same manner. Can you please let me know, if I need to correct something here?
> > Thanks, Suresh.
> 
> Sorry, unfortunately this PR won't be merged if CI isn't green. Are you sure that the `MSS_QSPI_INIT` is used at all? I.e., how does this file look like after preprocessing. Because it looks like the `DT_INST_FOREACH_STATUS_OKAY` doesn't call it at all, and that's why we get these unused symbols. Could you check that?

Hi @fkokosinski ,
I will check and get back to you.
We have tested it on Polarfire SOC Icicle board and we are able to see that "DT_INST_FOREACH_STATUS_OKAY" is getting called.
Thanks for your time.

**Comment by nagasureshkumar** [comment]
> > > Can we ignore this error? As this warning is treated as error in this test. The other SPI drivers are also modeled in the same manner. Can you please let me know, if I need to correct something here?
> > > Thanks, Suresh.
> > 
> > 
> > Sorry, unfortunately this PR won't be merged if CI isn't green. Are you sure that the `MSS_QSPI_INIT` is used at all? I.e., how does this file look like after preprocessing. Because it looks like the `DT_INST_FOREACH_STATUS_OKAY` doesn't call it at all, and that's why we get these unused symbols. Could you check that?
> 
> Hi @fkokosinski , I will check and get back to you. We have tested it on Polarfire SOC Icicle board and we are able to see that "DT_INST_FOREACH_STATUS_OKAY" is getting called. Thanks for your time.

Hi @fkokosinski ,

It seems that this (spi_mpfs_mss_qspi.c) driver is getting compiled for other architectures.
As the device tree files is not getting included but the driver is compiled, hence this error
I have to update the spi driver Kconfig file ( Kconfig.mchp_mss_qspi) to depend on SOC_MPFS, so that it will get compiled only for SOC_MPFS.
In my local setup I will only compile it for SOC_MPFS, hence I didn't see this warning.
I will update it.

Thanks,
Suresh

**Review by fkokosinski — DISMISSED** [review]
LGTM

File: drivers/spi/spi_mchp_mss_qspi.c [hunk 1]
```diff
@@ -0,0 +1,587 @@
...
+static int mss_qspi_configure(const struct device *dev,
+			      const struct spi_config *spi_cfg)
+{
+	const struct mss_qspi_config *cfg = dev->config;
+
+	if (spi_cfg->operation & (SPI_TRANSFER_LSB | SPI_OP_MODE_SLAVE
+		 | SPI_MODE_LOOP) || ((IS_ENABLED(CONFIG_SPI_EXTENDED_MODES) &&
+		 (spi_cfg->operation & (SPI_LINES_DUAL |
+			SPI_LINES_QUAD |
+		       SPI_LINES_OCTAL))))) {
```
[the remark this answers is not in the document]
Comment (nagasureshkumar) [reply drivers/spi/spi_mchp_mss_qspi.c:466]: Hi @edersondisouza,

Thanks for the review.
Do you want to split it something like below? for each condition?
if  (spi_cfg->operation &  SPI_OP_MODE_SLAVE) {
 LOG_ERR("Slave mode operation is not supported yet \n\r");
}
I can add it. 
Also do you see any other issues or have any other comments on this driver patch?

Thanks,
Naga Suresh

[the remark this answers is not in the document]
Comment (edersondisouza) [reply drivers/spi/spi_mchp_mss_qspi.c:466]: Something along these lines would work - with the the added benefit of better messages. 
And this is my only comment =D

[the remark this answers is not in the document]
Comment (nagasureshkumar) [reply drivers/spi/spi_mchp_mss_qspi.c:466]: Ok, I will update it.

Thanks,
Naga Suresh

**Review by fkokosinski — APPROVED** [review]

**Review by edersondisouza — APPROVED** [review]

**Pull request merged**